src/coreclr/jit/valuenum.cpp

```diff
@@ -4422,12 +4422,22 @@ bool ValueNumStore::VNEvalCanFoldBinaryFunc(var_types type, VNFunc func, ValueNu
             case GT_ROL:
             case GT_ROR:
 
-            case GT_EQ:
-            case GT_NE:
             case GT_GT:
             case GT_GE:
             case GT_LT:
             case GT_LE:
+                if (IsVNHandle(arg0VN) || IsVNHandle(arg1VN))
```

Comment:
Is this going to miss comparisons against 0/null

Reply:
Ah, that's a valid point. Do we need to handle it for GT/GE/LT/LE or only EQ/NE?

Reply:
@EgorBo may have a better idea. I'd expect we see it for all cases from time to time

Reply:
It should be easy enough to fix. I'll push it once I verify it doesn't break on the simple tests.

Reply:
It's a bit ugly to read... perhaps a helper `IsVNHandleOrNull` would be in order?

Reply:
Can we just conservatively put these checks into the beginning of `VNEvalCanFoldBinaryFunc` ? to only allow handles in pair with nullptr. Then we can check the diffs and decide whether it's worth it to relax it.

Reply:
Although, Comparisons are fine when both args are handles. E.g. when we compare two frozen objects